Source/Tools/FEXLoader/LinuxSyscalls/Syscalls.cpp

```diff
@@ -808,17 +830,16 @@ uint64_t UnimplementedSyscallSafe(FEXCore::Core::CpuStateFrame *Frame, uint64_t
 }
 
 void SyscallHandler::LockBeforeFork() {
-  // XXX shared_mutex has issues with locking and forks
-  // VMATracking.Mutex.lock();
-
-  // Add other mutexes here
+  VMATracking.Mutex.lock();
 }
 
-void SyscallHandler::UnlockAfterFork() {
-  // Add other mutexes here
-
-  // XXX shared_mutex has issues with locking and forks
-  // VMATracking.Mutex.unlock();
+void SyscallHandler::UnlockAfterFork(bool Child) {
+  if (Child) {
+    VMATracking.Mutex.DefaultInitialize();
```

Comment:
Due to fork semantics, the child also technically ends up owning the mutex but with a different PID with the internal TID tracking of the pthread mutex.

Reply:
I phrased the question poorly and actually meant to ask about the *active thread* of the child process (i.e. the only child process thread that may still own a "valid" lock).

Sounds like your reply about internal TID tracking still applies (so we can't safely call `std::mutex::unlock`?), but won't the thread later try to `unlock` the thread regardless (in the function that originally acquired the mutex)?

Reply:
The child process of fork can't safely call std::mutex::unlock, nor can it call the pthread equivalents. Once it gets initialized back to its default state with `StealAndDropActiveLocks` it is then safe to lock and unlock as it wants.

The parent thread of the fork process is still allows to lock and unlock as it pleases, and we lock all critical mutexes going in to a fork so that the rest of the code doing state tracking doesn't get in to a weird state.

Reply:
> The child process of fork can't safely call std::mutex::unlock, nor can it call the pthread equivalents. Once it gets initialized back to its default state with StealAndDropActiveLocks it is then safe to lock and unlock as it wants.

It *is* safe to lock+unlock, but is it safe to *just* unlock after the fork? From the looks of it the new code would fire an assertion in that case (as would `std::mutex` previously).

The scenario I'm worried about is where a fork happens from a thread that FEX holds an `std::unique_lock` in. Sure, `StealAndDropActiveLocks` will reset the underlying mutex, but the `unique_lock` destructor will still attempt to `unlock()` that mutex after its reset.

Reply:
> is it safe to just unlock after the fork? 
Depends on if it is the parent thread or the child process
Parent thread is fine, child process needs to `StealAndDropActiveLocks`.

> but the unique_lock destructor
These are all locked without using `std::unique_lock` and instead locking the mutex objects directly, there is no destructor being called here.

Reply:
> Depends on if it is the parent thread or the child process

I'm not concerned about the parent thread, all of this is referring to the child process.

>  Parent thread is fine, child process needs to StealAndDropActiveLocks.

All this is referring to StealAndDropActiveLocks already having been called after fork().

> These are all locked without using std::unique_lock and instead locking the mutex objects directly, there is no destructor being called here.

Instead of a `unique_lock` we'll have a manual `lock` + `unlock` pair in the acquiring the code then. Since we're talking about the active thread, we'll end up with a sequence of events like this:
1. lock in some FEX function F
2. fork (maybe due to a signal handler?) and then return to FEX function F
3. unlock at the end of FEX function F

Reply:
Actually, I think I'm hitting this point:
> While this likely doesn't hit all of the FEX mutexes, this hits the ones that are burning fires and are happening frequently.

Since we're only dealing with 3 specific mutexes, the issue pointed out above is easily avoided with manual care. More complicated scenarios (where we `fork` with an unknown set of mutexes being held in the forking thread) just don't happen, right?

Reply:
There should be no case in which we have locked a mutex and then forked (On the calling thread that is forking at least).
Fork being a syscall means it always comes from the the JIT where that is guaranteed to not hold any host side mutexes.
Any case where we would get interrupted while holding a critical mutex either disables the signal mask entirely, or ends up in the deferred signal path, which gets deferred to a point where a critical mutex isn't getting held.